workos/resources/list.py

```diff
@@ -178,3 +197,35 @@ def auto_paging_iter(self) -> Iterator[ListableResource]:
                     return
             yield self.data[index]
             index += 1
+
+
+class AsyncWorkOsListResource(
+    BaseWorkOsListResource,
+    Generic[ListableResource, ListAndFilterParams],
+):
+    async def auto_paging_iter(self) -> AsyncIterator[ListableResource]:
+        next_page: WorkOsListResource[ListableResource, ListAndFilterParams]
+        after = self.list_metadata.after
+        fixed_pagination_params, filter_params = self._parse_params()
+        index: int = 0
+
+        while True:
+            if index >= len(self.data):
+                if after is not None:
+                    next_page = await self.list_method(
+                        after=after, **fixed_pagination_params, **filter_params
+                    )
+                    self.data = next_page.data
+                    after = next_page.list_metadata.after
+                    index = 0
+                    continue
+                else:
+                    return
+            yield self.data[index]
+            index += 1
+
+
+SyncOrAsyncListResource = Union[
+    Awaitable[AsyncWorkOsListResource],
+    WorkOsListResource,
+]
```

Comment:
So this is FUN. I wanted this to be:
```
SyncOrAsyncListResource = Union[
    Awaitable[AsyncWorkOsListResource[ListableResource, ListAndFilterParams]],
    WorkOsListResource[ListableResource, ListAndFilterParams],
]
```
but Unions cannot be used as generics. This means I had to make the `Protocol` return type `SyncOrAsyncListResource` for list methods, instead of something like `SyncOrAsyncListResource[Directory, DirectoryFilterParams]`.

Reply:
I think to do something similar for events I'm creating a union of all event types and using that as a `bounds=` constraint on a type variable.... I don't know if that's helpful at all.

Reply:
Let me give that a go.